crates/ruff_python_parser/src/lib.rs

```diff
@@ -281,224 +214,199 @@ pub fn parse_expression_starts_at(source: &str, offset: TextSize) -> Result<Expr
-///
-/// This struct is created by the [`Tokens::kinds`] method.
-#[derive(Clone, Default)]
-pub struct TokenKindIter<'a> {
-    inner: std::iter::Flatten<std::slice::Iter<'a, LexResult>>,
-}
+impl Program<Mod> {
+    /// Attempts to convert the [`Program<Mod>`] into a [`Program<ModModule>`].
+    ///
+    /// This method checks if the `syntax` field of the program is a [`Mod::Module`]. If it is, the
+    /// method returns [`Some(Program<ModModule>)`] with the contained module. Otherwise, it
+    /// returns [`None`].
+    ///
+    /// [`Some(Program<ModModule>)`]: Some
+    fn try_into_module(self) -> Option<Program<ModModule>> {
+        match self.syntax {
+            Mod::Module(module) => Some(Program {
+                syntax: module,
+                tokens: self.tokens,
+                errors: self.errors,
+                comment_ranges: self.comment_ranges,
+            }),
+            Mod::Expression(_) => None,
+        }
+    }
 
-impl<'a> TokenKindIter<'a> {
-    /// Create a new iterator from a slice of [`LexResult`].
-    pub fn new(tokens: &'a [LexResult]) -> Self {
-        Self {
-            inner: tokens.iter().flatten(),
+    /// Attempts to convert the [`Program<Mod>`] into a [`Program<ModExpression>`].
+    ///
+    /// This method checks if the `syntax` field of the program is a [`Mod::Expression`]. If it is,
+    /// the method returns [`Some(Program<ModExpression>)`] with the contained expression.
+    /// Otherwise, it returns [`None`].
+    ///
+    /// [`Some(Program<ModExpression>)`]: Some
+    fn try_into_expression(self) -> Option<Program<ModExpression>> {
+        match self.syntax {
+            Mod::Module(_) => None,
+            Mod::Expression(expression) => Some(Program {
+                syntax: expression,
+                tokens: self.tokens,
+                errors: self.errors,
+                comment_ranges: self.comment_ranges,
+            }),
         }
     }
+}
 
-    /// Return the next value without advancing the iterator.
-    pub fn peek(&mut self) -> Option<(TokenKind, TextRange)> {
-        self.clone().next()
+impl Program<ModModule> {
+    /// Returns the module body contained in this program as a [`Suite`].
+    pub fn body(&self) -> &Suite {
+        &self.syntax.body
     }
 }
 
-impl Iterator for TokenKindIter<'_> {
-    type Item = (TokenKind, TextRange);
+impl Program<ModExpression> {
+    /// Returns the expression contained in this program.
+    pub fn expr(&self) -> &Expr {
+        &self.syntax.body
+    }
 
-    fn next(&mut self) -> Option<Self::Item> {
-        let &(ref tok, range) = self.inner.next()?;
-        Some((TokenKind::from_token(tok), range))
+    /// Consumes the [`Program`] and returns the parsed [`Expr`].
+    pub fn into_expr(self) -> Expr {
+        *self.syntax.body
     }
 }
 
-impl FusedIterator for TokenKindIter<'_> {}
+/// Tokens represents a vector of lexed [`Token`].
+#[derive(Debug)]
+pub struct Tokens {
+    raw: Vec<Token>,
```

Comment:
The `Tokens` struct will gain new methods as I understand how the lexer APIs are being used by the downstream tools but two APIs have been added:
1. `up_to_first_unknown` - Slice of tokens up to (and excluding) the first unknown token. This is what the linter sees.
2. `tokens_in_range` - This will replace most usages of `lex_starts_at`